### PR TITLE
test: add e2e test scenario for Kind extension

### LINKED ISSUE #8053: Deploy a pod to a Kubernetes (kind cluster)
Implement an End-to-End (E2E) test to verify the deployment of a pod/container in a Kubernetes cluster using the "Deploy to Kubernetes" option.
 
Steps to Implement the E2E Test:

- Creating of a Kind cluster (with ingress controller).
- Verification of the created cluster via podman desktop and basic operations -> restart, stop, start, delete.
- Check for resources added with kind cluster (images, containers, volumes, etc).
- Switching to newly created kube-context.
- Deploying a pod/container for a Kubernetes cluster via Deploy To Kubernetes option.
- Verification of the deployed app.

### PR DESCRIPTION
**Pull request by amisskii**
Signed-off-by: Anton Misskii <amisskii@redhat.com>

### What does this PR do?

This PR introduces an end-to-end test scenario for the Kind extension. 
The scenario involves: 
1. **Kind Cluster Creation**
   - Sets up a Kind cluster for testing purposes.

2. **Resource Validation**
   - Verifies that resources have been correctly added to the Kind cluster.

3. **Kubeconfig Context Validation**
   - Ensures that the kubeconfig context points to the newly created Kind cluster.

4. **Cluster Operations**

### What issues does this PR fix or reference?
#5976 #8053 
https://github.com/containers/podman-desktop/pull/8152

### How to test this PR?

Run the tests with yarn test:e2e:run

### CODE REVIEW

File: tests/playwright/src/model/pages/pods-details-page.ts [hunk 1]
```diff
@@ -61,7 +61,7 @@ export class PodDetailsPage extends BasePage {
   }
 
   async getState(): Promise<string> {
-    const currentState = await this.page.getByRole('status').getAttribute('title');
+    const currentState = await this.page.getByRole('status').nth(0).getAttribute('title');
```
Comment (amisskii) [inline tests/playwright/src/model/pages/pods-details-page.ts]: Maybe we should change the HTML code of the pod details page by defining new attributes for the parent area where the pod status and buttons are located, instead of using the nth() function?

File: tests/playwright/src/specs/kind.spec.ts [hunk 2]
```diff
@@ -54,34 +62,102 @@ beforeEach<RunnerTestContext>(async ctx => {
...
+      await kindExtension.disableExtension();
+      await navigationBar.openSettings();
+      await playExpect(resourcesPage.kindResources).not.toBeVisible();
+      await navigationBar.openExtensions();
+      await kindExtension.enableExtension();
+      await navigationBar.openSettings();
+      await playExpect(resourcesPage.kindResources).toBeVisible();
+    });
   });
+  describe.skipIf(process.env.GITHUB_ACTIONS && process.env.RUNNER_OS === 'Linux')(
```
Comment (amisskii) [inline tests/playwright/src/specs/kind.spec.ts:102]: Create_Kind_cluster_failure
There is an error with creating a Kind cluster using the Podman Desktop app on a Linux virtual machine.

File: tests/playwright/src/specs/kind.spec.ts [hunk 3]
```diff
@@ -54,34 +62,102 @@ beforeEach<RunnerTestContext>(async ctx => {
...
+
+        const volumesPage = await navigationBar.openVolumes();
+        await playExpect.poll(async () => await volumesPage.kindClusterVolumeExists(), { timeout: 10000 }).toBeTruthy();
+      });
+      test('Validate the right Kubernetes context is chosen', async () => {
+        await statusBar.validateKubernetesContext(kubernetesKontex);
+      });
+      test('Create a Pod to be deployed to the Kind cluster', async () => {
+        const podsPage = await navigationBar.openPods();
+        await createPod(podsPage, podAppName);
```
Comment (amisskii) [inline tests/playwright/src/specs/kind.spec.ts]: The https://github.com/containers/podman-desktop/pull/8152 must be merged first in order to use this function.

File: tests/playwright/src/model/pages/deploy-to-kubernetes-page.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import { BasePage } from '../pages/base-page';
+
+export class DeployToKubernetesPage extends BasePage {
+  readonly deployButton: Locator;
+  readonly doneButton: Locator;
+  readonly podNameField: Locator;
+  constructor(page: Page) {
+    super(page);
+    this.podNameField = this.page.locator('#podName');
```
Comment (amisskii) [inline tests/playwright/src/model/pages/deploy-to-kubernetes-page.ts]: We need to add attribute for this locator.

File: tests/playwright/src/model/pages/pods-details-page.ts (hunk 1 above)
Comment (cbr7) [reply tests/playwright/src/model/pages/pods-details-page.ts]: I'm not sure I understand why you needed to add the nth(0) param, is there another cell in that row with the role of `status`? The methods that are calling this element seem to be working fine.

File: tests/playwright/src/model/pages/pods-page.ts [hunk 5]
```diff
@@ -80,4 +85,11 @@ export class PodsPage extends MainPage {
     }
     return row.getByRole('button', { name: 'kebab menu', exact: true });
   }
+
+  public async openDeployToKubernetePage(name: string): Promise<DeployToKubernetesPage> {
+    const podActionMenu = await this.getPodActionsMenu(name);
+    await podActionMenu.click();
+    await this.deployToKubernetesButton.click();
```
Comment (cbr7) [inline tests/playwright/src/model/pages/pods-page.ts]: Should await for button to be enabled before clicking it.

File: tests/playwright/src/model/pages/resources-page.ts [hunk 6]
```diff
@@ -34,5 +40,26 @@ export class ResourcesPage extends SettingsPage {
...
+
+  public async kindClusterExists(label = 'kind-cluster'): Promise<boolean> {
+    return (await this.kindProviderConnections.getByRole('region', { name: label }).count()) > 0;
+  }
+
+  public async operateWithKindCluster(operation: string): Promise<void> {
+    await playExpect(
+      this.kindProviderConnections.getByRole('button', { name: operation, exact: true }),
+    ).not.toBeDisabled({ timeout: 7000 });
+    await this.kindProviderConnections.getByRole('button', { name: operation, exact: true }).click();
```
Comment (cbr7) [inline tests/playwright/src/model/pages/resources-page.ts]: If you are going to implement such a general method the param for it should be a value from an enum, not just a plain string, otherwise users won't know what possible values there are.

File: tests/playwright/src/model/pages/pods-details-page.ts (hunk 1 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/pods-details-page.ts]: I encountered an error while testing on a Windows machine. There was another element with the role of status.

File: tests/playwright/src/model/pages/volumes-page.ts [hunk 7]
```diff
@@ -36,6 +36,12 @@ export class VolumesPage extends MainPage {
     this.collectUsageDataButton = this.additionalActions.getByRole('button', { name: 'Gather volume sizes' });
   }
 
+  async kindClusterVolumeExists(): Promise<boolean> {
+    const volumesTable = await this.getTable();
+    const kindClusterVolume = volumesTable.getByRole('row').first();
+    return (await kindClusterVolume.count()) > 0;
+  }
+
```
Comment (cbr7) [inline tests/playwright/src/model/pages/volumes-page.ts]: Can you clarify on what this does? How is it different from just getting the first row in the volumes page?

File: tests/playwright/src/model/pages/pods-details-page.ts (hunk 1 above)
Comment (cbr7) [reply tests/playwright/src/model/pages/pods-details-page.ts]: Which element? There shouldnt be another element I don't think, perhaps you've encountered a bug in the app.

**Review by cbr7 — CHANGES_REQUESTED** [review]
The smoke test execution for this branch just exceeded 50 mins, I'm not sure on what is happened, but normal runtime is like 6-7 minutes.

https://github.com/containers/podman-desktop/actions/runs/10007792370/job/27663212571?pr=8151

File: tests/playwright/src/model/pages/volumes-page.ts (hunk 7 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/volumes-page.ts]: All available methods require at least one parameter, which is the volume name. Since a volume created with the cluster has a different name each time, I need a method that can check for the existence of any volume. I also attempted to extract volume names from the container's inspect data but I didn't get it.

Comment (cbr7) [reply tests/playwright/src/model/pages/volumes-page.ts]: ok, but how do you know the first volume is the volume for the kind cluster? What if the kind cluster volume is 2nd or 5th in the list?

File: tests/playwright/src/model/pages/pods-details-page.ts (hunk 1 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/pods-details-page.ts]: Screenshot from 2024-07-19 15-51-03

Comment (cbr7) [reply tests/playwright/src/model/pages/pods-details-page.ts]: Which button is this "Tasks"?

Comment (amisskii) [reply tests/playwright/src/model/pages/pods-details-page.ts]: I suppose that it is a bug, there isn't any 'Tasks' button on the test screenshots.

File: tests/playwright/src/model/pages/create-cluster-page.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    this.clusterHttpPort = this.clusterPropertiesInformation.getByLabel('HTTP Port', { exact: true });
+    this.clusterHttpsPort = this.clusterPropertiesInformation.getByLabel('HTTPS Port', { exact: true });
+    this.clusterIngressControllerCheckbox = this.clusterPropertiesInformation.getByRole('checkbox');
+    this.clusterCreationButton = this.clusterPropertiesInformation.getByRole('button', { name: 'Create', exact: true });
+    this.goBackButton = this.page.getByRole('button', { name: 'Go back to resources' });
+    this.logsButton = this.page.getByRole('button', { name: 'Show Logs' });
+  }
+
+  public async checkClusterProperties(): Promise<void> {
+    await playExpect(this.clusterPropertiesInformation).toBeVisible();
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-cluster-page.ts]: @amisskii  I am not sure if we need this kind of check for every element. We gonna check for some once we ie. expect the page to appear after create New button was pressed, then once we gonna start filling some info into fields, etc. whyt @cbr7 ?

File: tests/playwright/src/model/pages/deploy-to-kubernetes-page.ts (hunk 4 above)
Comment (odockal) [reply tests/playwright/src/model/pages/deploy-to-kubernetes-page.ts]: I would say we need to add more than just one things here. Whole page could make use of some ARIA functionality revision.

File: tests/playwright/src/model/pages/pods-details-page.ts (hunk 1 above)
Comment (odockal) [reply tests/playwright/src/model/pages/pods-details-page.ts]: Details page needs to get some ARIA functionality - #5966

File: tests/playwright/src/model/pages/resources-page.ts [hunk 9]
```diff
@@ -26,6 +29,10 @@ export class ResourcesPage extends SettingsPage {
   readonly podmanResources: Locator;
   readonly composeResources: Locator;
   readonly kindResources: Locator;
+  readonly kindProviderSetup: Locator;
```
Comment (odockal) [inline tests/playwright/src/model/pages/resources-page.ts]: I would say this does not belong here. Every "Provider connection card" on the resources page is generated dynamically. We should also approach it this way, thus, let's create an abstract/general class to cover shared functionality and behavior of the every resource connection card, and then we can add specific provider card  implementations if that differs. Something similar as in Extensions.

File: tests/playwright/src/specs/kind.spec.ts [hunk 10]
```diff
@@ -54,34 +68,107 @@ beforeEach<RunnerTestContext>(async ctx => {
...
+
+        const volumesPage = await navigationBar.openVolumes();
+        await playExpect
+          .poll(async () => await volumesPage.kindClusterVolumeExists(kindContainerName), { timeout: 10000 })
+          .toBeTruthy();
+      });
+      test('Validate the right Kubernetes context is chosen', async () => {
+        await statusBar.validateKubernetesContext(kubernetesKontex);
+      });
+      test('Create a Pod to be deployed to the Kind cluster', async () => {
```
Comment (odockal) [inline tests/playwright/src/specs/kind.spec.ts]: I would stop here, this could be added in another PR, otherwise this PR will grow too much.

**Review by odockal — CHANGES_REQUESTED** [review]

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  readonly clusterIngressControllerCheckbox: Locator;
+  readonly clusterCreationButton: Locator;
+  readonly goBackButton: Locator;
+  readonly logsButton: Locator;
+
+  constructor(page: Page) {
+    super(page);
+    this.clusterPropertiesInformation = this.page.getByRole('form', { name: 'Properties Information' });
+    this.clusterNameField = this.clusterPropertiesInformation.getByLabel('Name');
+    this.clusterIngressControllerCheckbox = this.clusterPropertiesInformation.getByRole('checkbox');
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: please add accessible name into getByrole. And, please add a locator for a parent element, that would allow you to actually change the checkbox value. Known issue. This is a workaround

File: tests/playwright/src/model/workbench/status-bar.ts [hunk 12]
```diff
@@ -37,4 +39,10 @@ export class StatusBar extends BasePage {
     await handleConfirmationDialog(this.page, 'Kind');
     await handleConfirmationDialog(this.page, 'Kind', true, 'OK');
   }
+
+  public async validateKubernetesContext(context: string): Promise<void> {
+    await playExpect(this.kubernetesContextButton).toBeVisible();
+    const currentContex = await this.kubernetesContextButton.textContent();
```
Comment (odockal) [inline tests/playwright/src/model/workbench/status-bar.ts]: it might be enough to just:
`await playExpect(kubernetesContext).toHaveText('context')`.

File: tests/playwright/src/model/pages/volumes-page.ts [hunk 13]
```diff
@@ -36,6 +36,30 @@ export class VolumesPage extends MainPage {
     this.collectUsageDataButton = this.additionalActions.getByRole('button', { name: 'Gather volume sizes' });
   }
 
+  public async getKindClusterVolumeName(containerName: string): Promise<string | undefined> {
```
Comment (odockal) [inline tests/playwright/src/model/pages/volumes-page.ts]: please rename the function as it can be generally used.

File: tests/playwright/src/model/pages/resources-kind-page.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import { expect as playExpect, type Locator, type Page } from '@playwright/test';
+
+import type { KindClusterOperations } from '../core/kind-cluster-operations';
+import { CreateKindClusterPage } from './create-kind-cluster-page';
+import { ResourcesPage } from './resources-page';
+
+export class ResourcesKindCard extends ResourcesPage {
```
Comment (odockal) [inline tests/playwright/src/model/pages/resources-kind-page.ts]: you should not inherit from resourcesPage since you cannot really use the locator from there in here. Just use BasePage. example: `src/model/pages/extension-card-page`.

Comment (odockal) [reply tests/playwright/src/model/pages/resources-kind-page.ts]: And make it please universal, since any other connection type can be used, add parameter for constructor.

Comment (odockal) [reply tests/playwright/src/model/pages/resources-kind-page.ts]: Maybe, the best would be if you create a follow up issue, where you will refactor the Resource Card Pages. Ie.:
Abstract ResourceCardPage and two classes that inherits from it, ResourceConnectionCardPage (podman, kind, etc.) and CliResourceCardPage (kubectl, compose).

Comment (amisskii) [reply tests/playwright/src/model/pages/resources-kind-page.ts]: https://github.com/containers/podman-desktop/pull/8310

File: package.json [hunk 15]
```diff
@@ -50,7 +50,7 @@
     "test:unit": "npm run test:main && npm run test:preload && npm run test:preload-docker-extension && npm run test:preload-webview && npm run test:ui && npm run test:renderer && npm run test:scripts:stylesheet && npm run test:tools && npm run test:extensions",
     "test:e2e": "npm run test:e2e:build && npm run test:e2e:run",
     "test:e2e:build": "cross-env NODE_ENV=development MODE=development DEBUG=pw:browser npm run build",
-    "test:e2e:run": "xvfb-maybe --auto-servernum --server-args='-screen 0 1280x960x24' -- vitest run tests/playwright/src/specs/ --pool=threads --poolOptions.threads.singleThread --poolOptions.threads.isolate --no-file-parallelism",
+    "test:e2e:run": "xvfb-maybe --auto-servernum --server-args='-screen 0 1280x960x24' -- vitest run tests/playwright/src/specs/kind.spec.ts --pool=threads --poolOptions.threads.singleThread --poolOptions.threads.isolate --no-file-parallelism",
```
Comment (odockal) [inline package.json]: do not forget to remove the spec from filter.

**Comment by amisskii** [comment]
https://github.com/containers/podman-desktop-e2e/actions/runs/10384270373

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    await playExpect(this.controllerCheckbox).toBeVisible();
+    switch (operation) {
+      case CheckboxOperations.Check:
+        await this.checkboxHolder.check();
+        await playExpect(this.controllerCheckbox).toBeChecked();
+        break;
+      case CheckboxOperations.Uncheck:
+        await this.checkboxHolder.uncheck();
+        await playExpect(this.controllerCheckbox).not.toBeChecked();
+        break;
```
Comment (cbr7) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: You should check that the desired end state of the checkbox is not also the initial state of it, because otherwise checking and unchecking will not work as you want them. If you want to check the box you should validate that the box isnt iniitally checked before performing the check action, if it is checked initially you can just return out of the method as completed.

File: tests/playwright/src/model/pages/main-page.ts [hunk 17]
```diff
@@ -75,14 +75,18 @@ export abstract class MainPage extends BasePage {
     return await this.page.getByRole('row').first().isVisible();
   }
 
+  async getAllTableRows(): Promise<Array<Locator>> {
```
Comment (cbr7) [inline tests/playwright/src/model/pages/main-page.ts]: Why is this a promise when it getting a list of Locators? It can be simplified to return Locator[]

File: tests/playwright/src/model/pages/main-page.ts [hunk 18]
```diff
@@ -75,14 +75,18 @@ export abstract class MainPage extends BasePage {
...
+    const table = await this.getTable();
+    return table.getByRole('row').all();
```
Comment (cbr7) [inline tests/playwright/src/model/pages/main-page.ts]:
```suggestion
    return await this.getTable().getByRole('row').all();
```

File: tests/playwright/src/model/pages/volumes-page.ts [hunk 19]
```diff
@@ -38,6 +38,30 @@ export class VolumesPage extends MainPage {
...
+            }
+          }
+          await volumeDetails.backLink.click();
+        }
+      }
+    } catch (err) {
+      console.log(`Exception caught on ${this.title} page with message: ${err}`);
+    }
+  }
+
```
Comment (cbr7) [inline tests/playwright/src/model/pages/volumes-page.ts]: I'm honestly not sure what this method is doing, but it strikes me as a bad idea to do all those actions and navigations inside a for loop like that, can you please explain the flow of this method and why it should be like this? As far as I understand you are performing page navigation, multiple times in fact, inside a for loop.

File: tests/playwright/src/model/pages/main-page.ts (hunk 17 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/main-page.ts]: The return type of an async method must be the global Promise<T> type

File: tests/playwright/src/model/pages/volumes-page.ts (hunk 19 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/volumes-page.ts]: Of course! We need to find a volume that was added to the Kind cluster. However, the information indicating that the volume belongs to the Kind cluster is only in the summary tab of the volume details page. Therefore, I will go through all the volumes in the volumes table and check if the summary tab contains information about this volume being used by the Kind container.

Comment (cbr7) [reply tests/playwright/src/model/pages/volumes-page.ts]: In that case that should be logic you implement inside the test, not inside a method, since this validation is test specific. Such logic is very brittle and should only be used inside the well known conditions of a test, it should not be part of exposed functionality from a POM.

File: tests/playwright/src/model/pages/main-page.ts (hunk 17 above)
Comment (odockal) [reply tests/playwright/src/model/pages/main-page.ts]: where did you see that?

File: tests/playwright/src/model/pages/volumes-page.ts (hunk 19 above)
Comment (odockal) [reply tests/playwright/src/model/pages/volumes-page.ts]: You are trying to find an association of the container's volume in a volume Page. And to map it, you need to open volume's details summary page, If I am not mistaken. Unless we got that info in Volume Page rows directly, only way.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts (hunk 16 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: The first property of the check() method is: _Ensure that element is a checkbox or a radio input. If not, this method throws. If the element is already checked, this method returns immediately_. Therefore, I believe there is no need to assert whether the element is checked before calling this method.

Comment (amisskii) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: Same for uncheck()

Comment (cbr7) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: That is not what I said, I said that you should check the checkbox is not already checked/unchecked before check/unchecking yourself.

File: tests/playwright/src/model/pages/volumes-page.ts (hunk 19 above)
Comment (amisskii) [reply tests/playwright/src/model/pages/volumes-page.ts]: Yes, I’m trying to find it on the details summary page and then just return the volume name to assert if it is running (used). Regarding implementing this method directly in the test, I think it could be useful in other tests as well. So, for me, the POM object is a good place for that functionality.

Comment (cbr7) [reply tests/playwright/src/model/pages/volumes-page.ts]: It really isn't because you can't control how the method will be used in the future and it will require heavy refactoring to be made non brittle, additionally the functionality is not even standard for POM objects, since the method is performing repeated navigations outside through different POM pages.

> I think it could be useful in other tests as well.

That is exactly what I am concerned about, the fact that people might be tempted to use this method in other tests without understanding what it involves and accounting for the different navigations that it performs.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+    switch (operation) {
+      case CheckboxOperations.Check:
+        await this.checkboxHolder.check();
+        await playExpect(this.controllerCheckbox).toBeChecked();
+        break;
+      case CheckboxOperations.Uncheck:
+        await this.checkboxHolder.uncheck();
+        await playExpect(this.controllerCheckbox).not.toBeChecked();
+        break;
```
Comment (cbr7) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts:78]:
```suggestion
    switch (operation) {
      case CheckboxOperations.Check:
        await playExpect(this.controllerCheckbox).not.toBeChecked();
        await this.checkboxHolder.check();
        await playExpect(this.controllerCheckbox).toBeChecked();
        break;
      case CheckboxOperations.Uncheck:
        await playExpect(this.controllerCheckbox).toBeChecked();
        await this.checkboxHolder.uncheck();
        await playExpect(this.controllerCheckbox).not.toBeChecked();
        break;
```

File: tests/playwright/src/model/pages/volumes-page.ts (hunk 19 above)
Comment (odockal) [reply tests/playwright/src/model/pages/volumes-page.ts]: Let's move it then outside of pom. into ie. utilities/operations.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts (hunk 16 above)
Comment (odockal) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: I would divide the create kind cluster function into two. One the will create and verifu default values in the form, and custom one, which takes multiple parameters and set them into the form.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+    httpPort = '9090',
+    httpsPort = '9443',
+    disableIngressController = false,
+    containerImage = '',
+  }: KindClusterOptions = {}): Promise<void> {
+    await this.fillTextbox(this.clusterNameField, clusterName);
+
+    if (changeProvideType) {
+      await playExpect(this.providerType).toBeVisible();
+      await this.providerType.selectOption({ value: 'docker' });
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: Given the nature of the element -> select/drop down. We always can have multiple option defined by either index or better string. So the string is the right type here. You can choose to add a check whether such option exists and throw if it does not, but select passed string value.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+    await playExpect(this.providerType).toHaveValue('podman');
+    await playExpect(this.httpPort).toHaveValue('9090');
+    await playExpect(this.httpsPort).toHaveValue('9443');
+    await playExpect(this.controllerCheckbox).toBeChecked();
+    await playExpect(this.containerImage).toBeEmpty();
+    await this.createCluster();
+  }
+
+  public async createClusterParametrized({
+    clusterName = 'kind-cluster',
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: you could probably add an undefined as or type to each property, which would mean to go with what is default. Counts for every property here.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+  }: KindClusterOptions = {}): Promise<void> {
+    await this.fillTextbox(this.clusterNameField, clusterName);
+
+    if (changeProvideType) {
+      await playExpect(this.providerType).toBeVisible();
+      await this.providerType.selectOption({ value: 'docker' });
+      await playExpect(this.providerType).toHaveValue('docker');
+    }
+
+    await this.fillTextbox(this.httpPort, httpPort);
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: It would change a little here:
```
if (options.httpPort) {
   await this.fillTextbox(this.httpPort, httpPort);
}
```

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+    await this.fillTextbox(this.httpsPort, httpsPort);
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: same as above

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    await playExpect(this.containerImage).toBeEmpty();
+    await this.createCluster();
+  }
+
+  public async createClusterParametrized({
+    clusterName = 'kind-cluster',
+    changeProvideType = false,
+    httpPort = '9090',
+    httpsPort = '9443',
+    disableIngressController = false,
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: I would go for directly named property: `userIngressController`. Default is `true`.

File: tests/playwright/src/model/core/types.ts [hunk 26]
```diff
@@ -20,3 +20,12 @@ export interface ContainerInteractiveParams {
   interactive?: boolean;
   attachTerminal?: boolean;
 }
+
+export interface KindClusterOptions {
+  clusterName?: string;
+  changeProvideType?: boolean;
```
Comment (odockal) [inline tests/playwright/src/model/core/types.ts]: consider `string | undefined` types.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 27]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    await this.fillTextbox(this.containerImage, containerImage);
+    await this.createCluster();
+  }
+
+  private async createCluster(): Promise<void> {
+    await playExpect(this.clusterCreationButton).toBeVisible();
+    await this.clusterCreationButton.click();
+    await playExpect(this.goBackButton).toBeVisible({ timeout: 120000 });
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: timeout could be parametrized.

File: tests/playwright/src/model/pages/main-page.ts (hunk 17 above)
Comment (odockal) [inline tests/playwright/src/model/pages/main-page.ts]: why not: `Promise<Locator[]>` ?

File: tests/playwright/src/model/workbench/status-bar.ts [hunk 28]
```diff
@@ -16,19 +16,21 @@
...
 export class StatusBar extends BasePage {
   readonly kindInstallationButton: Locator;
+  readonly kubernetesContext: Locator;
 
   constructor(page: Page) {
     super(page);
     this.kindInstallationButton = this.page.getByTitle(
       'Kind not found on your system, click to download and install it',
     );
+    this.kubernetesContext = this.page.getByTitle('Current Kubernetes contex');
```
Comment (odockal) [inline tests/playwright/src/model/workbench/status-bar.ts]: There is a typo: `context`

File: tests/playwright/src/utility/operations.ts [hunk 29]
```diff
@@ -215,3 +217,23 @@ export async function deletePodmanMachine(page: Page, machineVisibleName: string
 export function checkForFailedTest(result: TaskResult, runner: PodmanDesktopRunner): void {
   if (result.errors && result.errors.length > 0) runner.setTestPassed(false);
 }
+
+export async function getVolumeNameForContainer(page: Page, containerName: string): Promise<string | undefined> {
+  const volumePage = new VolumesPage(page);
```
Comment (odockal) [inline tests/playwright/src/utility/operations.ts]: You must open the Volumes Page first. Look at other functions here.

File: tests/playwright/src/specs/kind.spec.ts [hunk 30]
```diff
@@ -64,33 +70,89 @@ afterAll(async () => {
...
+    });
   });
+  describe.skipIf(process.env.GITHUB_ACTIONS && process.env.RUNNER_OS === 'Linux')(
+    'Kind cluster creation',
+    async () => {
+      test('Create a Kind cluster', async () => {
+        await navigationBar.openSettings();
+        await playExpect.poll(async () => resourcesPage.resourceCardIsVisible(RESOURCE_NAME)).toBeTruthy();
+        await playExpect(kindResourceCard.markdownContent).toBeVisible();
+        const createButton = kindResourceCard.createButton;
```
Comment (odockal) [inline tests/playwright/src/specs/kind.spec.ts]: this assignment is not neccessary.

File: tests/playwright/src/specs/kind.spec.ts [hunk 31]
```diff
@@ -64,33 +70,89 @@ afterAll(async () => {
...
+      });
+      test('Validate the right Kubernetes context is chosen', async () => {
+        await statusBar.validateKubernetesContext(KUBERNETES_CONTEXT);
+      });
+      test('Verify Kind cluster basic operations', async () => {
+        await navigationBar.openSettings();
+        await playExpect(kindResourceCard.resourceElementConnectionStatus).toHaveText(ResourceElementState.Running);
+        await kindResourceCard.performConnectionAction(ResourceElementActions.Stop);
+        await playExpect(kindResourceCard.resourceElementConnectionStatus).toHaveText(ResourceElementState.Off, {
+          timeout: 25000,
```
Comment (odockal) [inline tests/playwright/src/specs/kind.spec.ts]: I think that this timeouts might not be enough on CI.

File: tests/playwright/src/specs/kind.spec.ts [hunk 32]
```diff
@@ -64,33 +70,89 @@ afterAll(async () => {
...
+        });
+        await kindResourceCard.performConnectionAction(ResourceElementActions.Stop);
+        await playExpect(kindResourceCard.resourceElementConnectionStatus).toHaveText(ResourceElementState.Off, {
+          timeout: 25000,
+        });
+        await kindResourceCard.performConnectionAction(ResourceElementActions.Delete);
+        await playExpect(kindResourceCard.markdownContent).toBeVisible({ timeout: 25000 });
+      });
+    },
```
Comment (odockal) [inline tests/playwright/src/specs/kind.spec.ts:173]: Also add delete test.

**Review by odockal — CHANGES_REQUESTED** [review]

File: tests/playwright/src/model/pages/resource-connection-card-page.ts [hunk 33]
```diff
@@ -39,7 +39,7 @@ export class ResourceConnectionCardPage extends ResourceCardPage {
 
   public async performConnectionAction(operation: ResourceElementActions): Promise<void> {
     const button = this.resourceElementConnectionActions.getByRole('button', { name: operation, exact: true });
-    await playExpect(button).toBeEnabled({ timeout: 10000 });
+    await playExpect(button).toBeEnabled({ timeout: 25000 });
```
Comment (ScrewTSW) [inline tests/playwright/src/model/pages/resource-connection-card-page.ts]: Could also maybe be parametrized

**Review by ScrewTSW — APPROVED** [review]
LGTM, just one comment from me

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 34]
```diff
@@ -0,0 +1,115 @@
...
+    await this.clusterCreationButton.click();
+    await playExpect(this.goBackButton).toBeVisible({ timeout: timeout });
+    await this.goBackButton.click();
+  }
+
+  private async fillTextbox(textbox: Locator, text: string): Promise<void> {
+    await playExpect(textbox).toBeVisible();
+    await textbox.fill(text);
+    await playExpect(textbox).toHaveValue(text);
+  }
```
Comment (amisskii) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: Last week, I was looking for an issue that I could work on and noticed that we have at least two issues related to the Kubernetes cluster. Perhaps we should move the functions for creating a Kind cluster and checking if the Kind CLI is installed into the operations? Or would it be better to do this in a separate PR? WDYT @cbr7  @odockal?

Comment (cbr7) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: different issues need different PRs, we should always make PRs as small as possible for better granularity when it comes to changes

Comment (odockal) [reply tests/playwright/src/model/pages/create-kind-cluster-page.ts]: Yes, I agree, if there is a functionality that will be shared eventually, do it as part of another PR.

File: tests/playwright/src/model/pages/create-kind-cluster-page.ts [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+      }
+
+      if (httpPort) {
+        await this.fillTextbox(this.httpPort, httpPort);
+      }
+      if (httpsPort) {
+        await this.fillTextbox(this.httpsPort, httpsPort);
+      }
+
+      if (!useIngressController) {
```
Comment (odockal) [inline tests/playwright/src/model/pages/create-kind-cluster-page.ts]: I would add else branch since if the default setup will change, passing the `true` value wont change anything. so please also add else part.

**Review by odockal — APPROVED** [review]

**Review by cbr7 — APPROVED** [review]
